mlflow/models/utils.py

```diff
@@ -856,6 +887,91 @@ def _is_scalar(x):
+        raise MlflowException(f"Expected dtype to be DataType, got {type(dtype).__name__}")
+    if not np.isscalar(data):
+        raise MlflowException(f"Expected data to be scalar, got {type(data).__name__}")
+    # Reuse logic in _enforce_mlflow_datatype for type conversion
+    pd_series = pd.Series(data)
+    try:
+        pd_series = _enforce_mlflow_datatype("", pd_series, dtype)
+    except MlflowException:
+        raise MlflowException(
+            f"Failed to enforce schema of data `{data}` with dtype `{dtype.name}`"
+        )
+    return pd_series[0]
+
+
+def _enforce_array(data: Any, arr: Array, required=True):
+    if not required and data is None:
+        return None
+
+    if not isinstance(data, (list, np.ndarray)):
+        raise MlflowException(f"Expected data to be list or numpy array, got {type(data).__name__}")
+
+    if isinstance(arr.dtype, DataType):
+        data_enforced = [_enforce_datatype(x, arr.dtype) for x in data]
+    elif isinstance(arr.dtype, Object):
+        data_enforced = [_enforce_object(x, arr.dtype) for x in data]
+    elif isinstance(arr.dtype, Array):
+        data_enforced = [_enforce_array(x, arr.dtype) for x in data]
+    else:
+        raise MlflowException(
+            f"Failed to enforce schema of data `{data}` with dtype `{arr}`. "
+            f"Invalid schema, `{arr.dtype}` is not supported type for Array element."
+        )
+
+    # Keep input data type
+    if isinstance(data, np.ndarray):
+        data_enforced = np.array(data_enforced)
+
+    return data_enforced
+
+
+def _enforce_property(data: Any, property: Property):
+    if isinstance(property.dtype, DataType):
+        return _enforce_datatype(data, property.dtype)
+    if isinstance(property.dtype, Array):
+        return _enforce_array(data, property.dtype)
+    if isinstance(property.dtype, Object):
+        return _enforce_object(data, property.dtype)
+    raise MlflowException(
+        f"Failed to enforce schema of data `{data}` with dtype `{property.dtype}`"
+    )
+
+
+def _enforce_object(data: Dict[str, Any], obj: Object, required=True):
+    if not required and data is None:
+        return None
+    if not isinstance(data, dict):
+        raise MlflowException(
+            f"Failed to enforce schema of '{data}' with type '{obj}'. "
+            f"Expected data to be dictionary, got {type(data).__name__}"
+        )
+    if not isinstance(obj, Object):
+        raise MlflowException(
+            f"Failed to enforce schema of '{data}' with type '{obj}'. "
+            f"Expected obj to be Object, got {type(obj).__name__}"
+        )
+    properties = {prop.name: prop for prop in obj.properties}
+    required_props = {k for k, prop in properties.items() if prop.required}
+    missing_props = required_props - set(data.keys())
+    if missing_props:
+        raise MlflowException(f"Missing required properties: {missing_props}")
+    invalid_props = data.keys() - properties.keys()
+    if invalid_props:
```

Comment:
```suggestion
    if invalid_props := data.keys() - properties.keys():
```

Reply:
Could we remove the addressed comments in code and update the nits in a PR against this branch?

Reply:
sounds good.